Reviews and refactors tests per specification

weierophinney opened the pull request:
Per the [published specification](https://docs.zendframework.com/zend-expressive/v3/features/container/config/), and based on my understanding of how zend-servicemanager is intended to operate, I've updated the proposed tests.

Some specifics:

- I updated test method names to better reflect the behavior being tested.
- I ensured that invokable tests were testing the correct behavior. In most cases, we want to test retrieving an invokable service, and _not_ the alias form (the alias tests test that aspect).
- I ensured delegators _do not act on items in the "services" configuration, nor on aliases of them_.
- I ensured delegators act _only on the resolved service_, and not per-alias.
- I expanded factory test cases to include named functions and invokable class _instances_.
- I renamed variables to reduce confusion with what we are testing. E.g., "factory", "delegator", and "invokable" as the return value for a `$container->get()` call were renamed to "instance".

You can look through the commit history to see all specifics, but I believe at this time, the tests reflect usage expectations.

zend-servicemanager _does_ pass at this time. The Aura.Di and Pimple tests do not, and appear to reflect the fact that we did not have a full specification previously!

webimpress commented on src/DelegatorTestTrait.php:
```diff
@@ -150,62 +190,107 @@ public function testDelegatorForAliasFactory() : void
+
+        // Now ensure that subsequent retrievals by alias retrieve the same
+        // instance.
+        self::assertSame($instance, $container->get('alias'));
```
Not sure if we need it here, that functionality should be tested in `AliasTestTrait`.

(delegator was not applied so this is basic service fetching from container)

webimpress commented on src/DelegatorTestTrait.php:
```diff
@@ -150,62 +190,107 @@ public function testDelegatorForAliasFactory() : void
+        self::assertNotInstanceOf(TestAsset\Delegator::class, $instance);
+
+        // Now ensure that the service fetched by alias is the same as that
+        // fetched by the canonical service name.
```
Also this comment should be updated, imho:
```
// Now ensure that the service fetched by the canonical service name
// is the same as that fetched by alias.
```

webimpress commented on src/DelegatorTestTrait.php:
```diff
@@ -150,62 +190,107 @@ public function testDelegatorForAliasFactory() : void
+        self::assertInstanceOf(TestAsset\Service::class, ($instance->callback)());
+
+        // Now ensure that the service fetched by alias is the same as that
+        // fetched by the canonical service name.
```
The same here about the comment.

webimpress commented on src/FactoryTestTrait.php:
```diff
@@ -16,24 +16,24 @@
+        yield 'invokable-instance'   => [['factories' => ['service' => new TestAsset\Factory()]]];
+        yield 'callable-array'       => [['factories' => ['service' => [TestAsset\FactoryStatic::class, 'create']]]];
+        yield 'callable-string'      => [['factories' => ['service' => TestAsset\FactoryStatic::class . '::create']]];
+        yield 'closure'   => [[
```
Lets move the second opening array bracket to the new line:
```
yield 'closure' => [
    [
        'factories' => [
            'service' => function () {
                return new TestAsset\Service();
            },
        ],
    ],
];
```

Only this is going to be valid with new zend-coding-standard (every value in multiline array must be in new line, and `[[` are not allowed when closing brackets are not in the same line).
I think it's a bit clearer.

webimpress commented on README.md:
```diff
@@ -43,4 +45,4 @@ following traits into your test case:
 - `Zend\ContainerConfigTest\ServiceTestTrait` - to support `services` configuration,
 
-or use `Zend\ContainerConfigTest\AllTestTrait` to support whole `zend-servicemanager` configuration.
+or use `Zend\ContainerConfigTest\AllTestTrait` to support the entire configuration.
```
Do we need this trait? Maybe we should rename class `ContainerTest` to `ExpressiveContainerTest` and include all these traits there, as we require that functionality for "expressive compatible container".

Then we can add (bring back) `SharedTestTrait` and it could be used separately with implementations which want support shared services.

And then, when we use the repository also for mvc project we can create class `MvcContainerTest` with all tests required for "mvc compatible container".

webimpress commented:
I can think of two more test cases:

1.
Config:
```php
$config = [
    'invokables' => [
        'alias1' => TestAsset\Service::class,
        'alias2' => TestAsset\Service::class,
    ],
];
```
Expected behaviour:
```php
self::assertSame($container->get('alias1'), $container->get('alias2'));
```

2.
Config:
```php
$config = [
    'invokables' => [
        'alias1' => TestAsset\Service::class,
        'alias2' => TestAsset\Service::class,
    ],
    'delegators' => [
        TestAsset\Service::class => [
            TestAsset\Delegator1Factory::class,
            TestAsset\Delegator2Factory::class,
        ],
    ],
];
```

Expected behaviour:
```php
self::assertTrue($container->has('alias1'));
self::assertTrue($container->has('alias2'));
$instance = $container->get('alias1');
self::assertInstanceOf(TestAsset\Service::class, $instance);
self::assertEquals(
    [
        TestAsset\Delegator1Factory::class,
        TestAsset\Delegator2Factory::class,
    ],
    $instance->injected
);

// Ensure subsequent retrievals get same instance
self::assertSame($instance, $container->get('alias1'));
self::assertSame($instance, $container->get('alias2'));
```

webimpress commented:
@weierophinney What about two more test cases:

1. Factory class exists but is not invokable
```php
$config = [
    'factories' => [
        'service' => TestAsset\NonIvokableFactory::class,
    ],
];
```

Expected behaviour:
```php
self::assertTrue($container->has('service'));

$this->expectException(\Psr\Container\ContainerExceptionInterface::class);
$container->get('service');
```

2. Factory class does not exist:

```php
$config = [
    'factories' => [
        'service' => TestAsset\FactoryDoesNotExist::class,
    ],
];
```

Expected behaviour:
```php
self::assertTrue($container->has('service'));

$this->expectException(\Psr\Container\ContainerExceptionInterface::class);
$container->get('service');
```

weierophinney commented on src/DelegatorTestTrait.php:
```diff
@@ -119,19 +153,25 @@ public function testDelegatorForAliasService() : void
+        self::assertSame($myService, $instance);
+
+        // Now ensure that the service fetched by alias is the same as that
+        // fetched by the canonical service name.
```
The main thing is that the two instances are the same; the order in which that is described is less relevant, as the test shows that clearly.

> BTW: is `foo-bar` canonical service name

Yes, in this case `foo-bar` is the resolved service name. The code previous to what's quoted above clearly demonstrates fetching by the alias name (here, `alias`).

Maybe I'm not understanding your question/assertion here, though...?

webimpress replied:
> The main thing is that the two instances are the same; the order in which that is described is less relevant, as the test shows that clearly.

Yes, I'd like just update the comment to say:
```php 
// Now ensure that the service fetched by the canonical service name
// is the same as that fetched by alias.
```

because (as I said before): we already know that "service fetched by alias is right" (assertions above) and this instance is in "expected" parameter (first param) - so we fetching service by the canonical service name (and this is what we have as 2nd param - "actual").

weierophinney commented on src/DelegatorTestTrait.php:
```diff
@@ -119,19 +153,25 @@ public function testDelegatorForAliasService() : void
     }
 
-    public function testDelegatorForAliasFactory() : void
+    public function testDelegatorsDoNotTriggerForAliasName() : void
```
The point is that the delegators cannot be named for the alias, but must be named for the canonical (resolved) service name. That may be a factory **or** an invokable class name. I think the test name here is accurate as-is.

weierophinney commented:
@webimpress —

I have done the following:

- Updated comment verbiage to make what we're testing more clear.
- Updated (future) CS issues, per your comments.
- Added tests as detailed in your comments above.
